src/mlir/consumer.h

```diff
@@ -0,0 +1,149 @@
+// Copyright Microsoft and Project Verona Contributors.
+// SPDX-License-Identifier: MIT
+
+#pragma once
+
+#include "generator.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "mlir/IR/MLIRContext.h"
+#include "parser/ast.h"
+
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/Error.h"
+
+#include <string>
+
+namespace mlir::verona
+{
+  /**
+   * Field offset maps between field names, types and their relative position.
+   *
+   * FIXME: This will need to be better encapsulated, but we'll expand it when
+   * we start implementing classes fully, and dynamic dispatch.
+   */
+  struct FieldOffset
+  {
+    llvm::SmallVector<llvm::StringRef, 1> fields;
+    llvm::SmallVector<Type, 1> types;
+  };
+
+  /**
+   * AST Consumer
+   */
+  class ASTConsumer
+  {
+    ASTConsumer(MLIRContext* context) : gen(context) {}
+
+    /// MLIR Generator
+    MLIRGenerator gen;
+
+    /// Map for each type which fields does it have.
+    std::map<llvm::StringRef, FieldOffset> classFields;
+
+    /// Function scope, for mangling names. 3 because there will always be the
+    /// root module, the current module and a class, at the very least.
```

Comment:
That sounds like a good reason for picking a number like 5.  The size for `llvm::Small*` should capture the most common case, not the minimum case.

Same comment as above about lifetimes.  I presume these StringRefs are things from higher up the AST than the code that we're currently parsing, but if that's an invariant on which memory safety depends then please document it.